### PR TITLE
New Adapter: Theadx

### PR DESCRIPTION
**Pull request by mustafakemal16**
Theadx server adaptor added.

### CODE REVIEW

File: adapters/theadx/theadx.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+	return bidResponse, errors
+}
+
+func getMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	if bid.Ext != nil {
+		var bidExt openrtb_ext.ExtBid
+		err := json.Unmarshal(bid.Ext, &bidExt)
+		if err == nil && bidExt.Prebid != nil {
+			return openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply adapters/theadx/theadx.go:141]: ^ Bad bot. This adapter is reading the bid type from the bidder's response.

File: exchange/adapter_builders.go [hunk 2]
```diff
@@ -398,5 +399,6 @@ func newAdapterBuilders() map[openrtb_ext.BidderName]adapters.Builder {
 		openrtb_ext.BidderZeroClickFraud:    zeroclickfraud.Builder,
 		openrtb_ext.BidderZetaGlobalSsp:     zeta_global_ssp.Builder,
 		openrtb_ext.BidderZmaticoo:          zmaticoo.Builder,
+		openrtb_ext.BidderTheadx:            theadx.Builder,
```
Comment (SyntaxNode) [inline exchange/adapter_builders.go:402]: Please add the adapter alphabetically.

File: openrtb_ext/bidders.go [hunk 3]
```diff
@@ -213,6 +213,7 @@ var coreBidderNames []BidderName = []BidderName{
 	BidderZeroClickFraud,
 	BidderZetaGlobalSsp,
 	BidderZmaticoo,
+	BidderTheadx,
```
Comment (SyntaxNode) [inline openrtb_ext/bidders.go:216]: Please add the adapter alphabetically.

File: openrtb_ext/bidders.go [hunk 4]
```diff
@@ -492,6 +493,7 @@ const (
 	BidderZeroClickFraud    BidderName = "zeroclickfraud"
 	BidderZetaGlobalSsp     BidderName = "zeta_global_ssp"
 	BidderZmaticoo          BidderName = "zmaticoo"
+	BidderTheadx            BidderName = "theadx"
```
Comment (SyntaxNode) [inline openrtb_ext/bidders.go:496]: Please add the adapter alphabetically.

Comment (mustafakemal16) [reply openrtb_ext/bidders.go]: theadx  moved to alphabeticall order.

**Review by mustafakemal16 — COMMENTED** [review]
theadx moved to alphabeticall order.

File: adapters/theadx/theadx.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mustafakemal16) [reply adapters/theadx/theadx.go:141]: its ok,it is used for multiple format/media support.

[the remark this answers is not in the document]
Comment (mustafakemal16) [reply adapters/theadx/theadx.go:141]: it is used for multiple format/media support.

File: openrtb_ext/bidders.go (hunk 3 above)
Comment (mustafakemal16) [reply openrtb_ext/bidders.go]: moved to alphabeticall order

**Review by mustafakemal16 — COMMENTED** [review]
all changes complete

**Review by mustafakemal16 — COMMENTED** [review]
all changes complete

File: static/bidder-info/theadx.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+endpoint: "https://ssp.theadx.com/request?pbs=1"
+maintainer:
+  email: "ssp@theadx.com"
```
Comment (Sonali-More-Xandr) [inline static/bidder-info/theadx.yaml:3]: @mustafakemal16 We have sent a verification email to ssp@theadx.com. Please reply to that email with 'Received' to complete the verification process. Thank you!

File: adapters/theadx/theadx.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
```
Comment (onkarvhanumante) [inline adapters/theadx/theadx.go:104]: https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L9-L28

just a suggestion. could make use of these response utils

File: static/bidder-info/theadx.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+endpoint: "https://ssp.theadx.com/request?pbs=1"
```
Comment (onkarvhanumante) [inline static/bidder-info/theadx.yaml:1]: endpoint is reachable
```
curl -i --location --request POST 'https://ssp.theadx.com/request?pbs=1'
HTTP/2 204
server: nginx/1.18.0 (Ubuntu)
```

File: static/bidder-info/theadx.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+      - video
+  site:
+    mediaTypes:
+      - banner
+      - native
+      - video
+userSync:
+  redirect:
+    url: "https://ssp.theadx.com/cookie?redirect_url={{.RedirectURL}}&?pbs=1"
+    userMacro: "$UID"
```
Comment (onkarvhanumante) [inline static/bidder-info/theadx.yaml:21]: @mustafakemal16 redirect url is not working

```
https://ssp.theadx.com/cookie?redirect_url=https%3A%2F%2Fib.adnxs.com%2Fprebid%2Fsetuid%3Fbidder%3Dtheadx%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%22%24UID%22&?pbs=1
```

**Comment by onkarvhanumante** [comment]
@mustafakemal16  should create docs PR in https://github.com/prebid/prebid.github.io repo. This PR is needed for include ResetDigital on prebid server docs - https://prebid.org/product-suite/prebid-server/

File: adapters/theadx/theadx.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mustafakemal16) [reply adapters/theadx/theadx.go:141]: audio added to static/bidder-info/theadx.yaml   file

**Comment by mustafakemal16** [comment]
> @mustafakemal16 should create docs PR in https://github.com/prebid/prebid.github.io repo. This PR is needed for include Theadx on prebid server docs - https://prebid.org/product-suite/prebid-server/

https://github.com/prebid/prebid.github.io/pull/5184

**Review by gargcreation1992 — APPROVED** [review]

**Review by Sonali-More-Xandr — APPROVED** [review]

**Pull request merged**